src/coreclr/jit/fgwasm.cpp

```diff
@@ -0,0 +1,813 @@
+// need to increase so the earlier interval can nest inside. That is, if have a:[0, 4] and b:[2,6] we
+// will need to emit them as b:[0,6], a[0,4].
+//
+// To save some time we also create a union-find like setup to identify the first interval in a set of
+// conflicting intervals. Say we have a:[0,4] b:[2,6] c:[5,7]. When we see that b conflicts with a,
+// we note 'a' as the conflict "chain" for b, and also track the conflict extent in a. Then when
+// we scan intervals for c, we see it conflicts with the chain starting at a, and we add it to the chain.
+// The net result is a:[0,4(7)], b:[2,6]-->a, c:[5,7]-->a.
+//
+// Then we order on their conflict chain start and end extent, and so would emit c:[0,7], b:[0,6], a:[0,4]
+//
+// We then can use the properly ordered and nested intervals to track the control stack depth as we
+// traverse the blocks in loop-aware RPO order, and emit the proper Wasm control flow.
+//
+// Still TODO
+//
+// * proper DFS/Loop finding
+// * handling irreducible loops
+// * handling funclets
+// * proper handling of BR_TABLE defaults
+// * branch inversion
+// * actual block reordering
+// * instruction emission
+// * tail calls (RETURN_CALL)
+// * UNREACHED in more places (eg noreturn calls)
+// * Rethink need for BB0 (have m_end refer to end of last block in range, not start of first block after)
+// * We do not branch with operands on the wasm stack, so we need to add suitable (void?) types to branches
+// * During LaRPO formation, remember the position of the last block in the loop
+//
+PhaseStatus Compiler::fgWasmControlFlow()
+{
+    // -----------------------------------------------
+    // (1) Build loop-aware RPO layout
+    //
+    if (m_dfsTree == nullptr)
+    {
+        m_dfsTree = fgComputeDfs</* useProfile */ true>();
+        m_loops   = FlowGraphNaturalLoops::Find(m_dfsTree);
+    }
+    else
+    {
+        assert(m_loops != nullptr);
+    }
+
+    // Bail out for now if there is any irreducible flow.
+    // TODO: run the irreducible flow fixing before this.
+    //
+    if (m_loops->ImproperLoopHeaders() > 0)
+    {
+        JITDUMP("\nThere are irreducible loops here, bailing\n");
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    // Our interval ends are at the starts of blocks, so we need a block that
+    // comes after all existing blocks. So allocate one extra slot.
+    //
+    JITDUMP("\nCreating loop-aware RPO\n");
+    BasicBlock** const initialLayout = new (this, CMK_Wasm) BasicBlock*[m_dfsTree->GetPostOrderCount() + 1];
+
+    // TODO: extend this to cover all the funclets as well.
+    //
+    // The "DFS" we run above should skip from CALLFINALLY to CALLFINALLYRET, treat all funclet returns
+    // as having no successor, and add each funclet entry as a DFS seed. This will give us a disjoint DFS
+    // tree for the main method and each funclet, and the layout below will properly transform them all into
+    // Wasm control flow.
+    //
+    unsigned numHotBlocks  = 0;
```

Comment:
Nit:
```suggestion
    unsigned numDfsBlocks  = 0;
```
The blocks we're missing here are the throw helper ones which represent implicit flow. The fact they're cold is not a key feature.